expression/constant.go

```diff
@@ -450,3 +451,66 @@ func (c *Constant) Coercibility() Coercibility {
 	}
 	return c.collationInfo.Coercibility()
 }
+
+// TryUpdateEQFunctionCollate Update collate based on priority #35149
+// Try to unify collate by priority before physical optimization.
+// After unifying collate, the impact of session level on some optimizations can be shielded
+func TryUpdateEQFunctionCollate(eq *ScalarFunction) {
+
+	if eq.FuncName.L == ast.EQ {
+		expressionUpdateCollate(eq)
+	}
+
+}
+
+// ExpressionUpdateCollate checks whether the expression can be updated to the new collation.
+func expressionUpdateCollate(eq *ScalarFunction) bool {
+	// If left and right are columns/Constant, they have the same priority so don't consider this case
+	var col *Column
+	var con *Constant
+	colIsLeft := true
+	colOk := false
+	conOk := false
+
+	col, colOk = eq.GetArgs()[0].(*Column)
+	if !colOk {
+		col, colOk = eq.GetArgs()[1].(*Column)
+		colIsLeft = false
+	}
+	if !colOk {
+		return false
+	}
+	if !col.HasCoercibility() {
+		return true
+	}
+
+	if colIsLeft {
+		con, conOk = eq.GetArgs()[1].(*Constant)
+	} else {
+		con, conOk = eq.GetArgs()[0].(*Constant)
+	}
+	if !conOk {
+		return false
+	}
+	if !con.HasCoercibility() {
+		return true
+	}
+
+	// In addition to utf8 and utf8mb4 encoding, other encodings are handled by the cast function
+	if col.GetType().GetCharset() != con.GetType().GetCharset() {
+		// utf8 is a subset of utf8mb4
+		// If the constant is utf8 and the column is utf8mb4, the constant can be directly converted to utf8mb4
+		if con.GetType().GetCharset() == "utf8" && col.GetType().GetCharset() == "utf8mb4" {
+			// The lower the coer value, the higher the priority
+			if col.Coercibility() < con.Coercibility() {
+				con.GetType().SetCharset(col.GetType().GetCharset())
+				con.GetType().SetCollate(col.GetType().GetCollate())
```

Comment:
Modify the expression may cause two problems:
1. data race
2. some place may referencing the expression and find  it's unexpectedly modified ... this kind is difficult to debug

"X eq Y" and after the expression execution, they are modified ... this is a footgun, you're playing with fire.